### ISSUE TITLE
Dynamic Snippet Form problem

### ISSUE THREAD

**Issue by diaafares** [tinkerun/tinkerun#17]
Hi,
At first, thanks for your efforts doing this great app.

Today I have an idea about to make Snippet Form dynamic, so I did something like this:
```php
use User\Models\User;

$users = User::selectRaw('first_name as label, id as value')->get()->toArray();

$field_user = [
    'value' => '',
    'label' => 'User',
    'description' => 'Select a user.',
    'type' => 'select',
    'options' => $users, 
];
```

But as soon as I click the Snippet Form button, I get this white empty screen:
![image](https://user-images.githubusercontent.com/1811607/120678543-deff0b00-c4a0-11eb-9162-4cdaddbd9e14.png)

I know maybe the feature was not design to work as I wanted, but when I think about it, supporting dynamic Snippet Form will make the possibilities endless for Tinkerun

**Comment by billyct** [comment]
yes, it not support the dynamic Snippet Form. your idea is so cool.

I will try to do the auto-update.

**Comment by billyct** [comment]
@diaafares do you have any other use case of dynamic Snippet Form?

**Comment by diaafares** [comment]
Hi @billyct 
I told you about my use case at the previous comment, that I want to create dynamic Snippet Form input that are populated from the Database, then use these inputs values as parameters for DB queries to get the final results.

For example: select a user to get how many post he have and other statistics.
The possibilities are endless, you can whatever you want if we get this feature.

I have an idea about how you may implement the feature:
maybe we can a prefix for the variables that we want to make them dynamic, like adding: `$dynamic_` to the variable `$dynamic_users` and when you click the Snippet Form button, Tinkerun will search for these variables that have `$dynamic_` prefix and run them in Tinker to get their values and cache it, after that, it should be able to render the Snippet Form with the right values.

Just wanted to give you an idea how you can implement this feature, but maybe you have a better solution for this ☺️

**Comment by billyct** [comment]
@diaafares I have done a simple auto-update for mac in [v0.2.4](https://github.com/tinkerun/tinkerun/releases/tag/v0.2.4)

**Comment by diaafares** [comment]
@billyct oh, great work, thank you very much
BTW, in the v0.2.4, I tried the dynamic Snippet Form example code above and it didn't give me an empty screen! instead it display the select input but the select was empty:
![image](https://user-images.githubusercontent.com/1811607/120891125-771b0280-c60f-11eb-8546-e649c3103a0c.png)

**Comment by billyct** [comment]
yes, I have not done the dynamic Snippet Form yet.

how about writing the dynamic data in the closure

for your case above:
```php
$field_user = [
    'value' => '',
    'label' => 'User',
    'description' => 'Select a user.',
    'type' => 'select',
    'options' => function() {
		return User::selectRaw('first_name as label, id as value')->get()->toArray();
	}, 
];
```

with the closure, so I can just run code in the closure only with tinker for getting the dynamic data.

**Comment by diaafares** [comment]
> how about writing the dynamic data in the closure
@billyct oh, great idea, way better than what I suggested!

It will be great if we can do it like this:
```php
$options = function() {
    return User::selectRaw('first_name as label, id as value')->get()->toArray();
};

$field_user = [
    'value' => '',
    'label' => 'User',
    'description' => 'Select a user.',
    'type' => 'select',
    'options' => $options, 
];
```
Is that possible? ☺️

**Comment by billyct** [comment]
<img src="https://user-images.githubusercontent.com/1612364/121814649-c367e180-cca4-11eb-9609-3c34d7aa805b.gif" alt="Dynamic Snippet Form" width="400"/>

😉 Already available in [v0.2.5](https://github.com/tinkerun/tinkerun/releases/tag/v0.2.5)

**Comment by diaafares** [comment]
@billyct awesome work, thank you very much ^_^
Please add it to the change log, I can't wait to tell my teammates about it ☺️

**Comment by diaafares** [comment]
@billyct BTW, I tried to get v0.2.5 by using auto update feature, but it seems like it didn't work, I opened Tinkerun and it was downloading the new version in the background, I waited until the network activity stopped (it should mean that the download finished) but nothing happened, I restarted the app and I still have v0.2.4.

**Comment by billyct** [comment]
@diaafares I will check the auto-updater.

And Tinkerun cannot handle 10000 records, because it is only simulating Tinker’s Terminal input.

if I do the 10000 records in the terminal, Tinker will just send me something like below:

```shell
App\Models\User {#5477 …6},
App\Models\User {#5478 …6},
App\Models\User {#5479 …6},
App\Models\User {#5480 …6},
```

I think you should add some conditions for your query 😂

**Comment by diaafares** [comment]
@billyct I added `limit(3)` to my query and I have the same error!
I think the cause of the problem is that when you did a query to table that have a large data, it will take some time to execute it even if the results are small.

Or maybe the problem is different than my thoughts I really don't know

**Comment by billyct** [comment]
@diaafares how about just run some code like below in tinker

```php
json_encode((function() {
    return User::selectRaw('first_name as label, id as value')->limit(3)->get()->toArray();
})())
```

I get the dynamic data in the background by this code.

**Comment by diaafares** [comment]
@billyct running your code using tinker from the terminal give this result (there are no errors): 
```
"[{"label":"Josh","value":2270,"name":"","initials":""},{"label":"Vladimir","value":2271,"name":"","initials":""},{"label":"Daniela","value":2272,"name":"","initials":""}]"
```

I don't know why the original code doesn't work inside Tinkerun app

**Comment by diaafares** [comment]
@billyct I think the problem is happen because of the import of the User model when Tinkerun execute the closures, as my App have custom setup and the User model is not placed as normal Laravel apps, I ran the code inside Tinker from terminal and it give me the following error:
```
Aliasing 'User' to 'User\Contracts\User' for this Tinker session.
PHP Error:  Call to undefined method User\Contracts\User::selectRaw() in /www/testappeval()'d code on line 2
```

The problem is I added the right import inside Tinkerun, but as I said before, it seems like Tinkerun doesn't recognize the imports inside the closures, what you think ☺️

**Comment by billyct** [comment]
@diaafares yes, Tinkerun can't recognize the things outside the closures.

so in your case, you should use 

```php
App\Models\User::selectRaw('first_name as label, id as value')->limit(3)->get()->toArray();
```

not good, but I think it will work 😂 

Because Tinkerun does not depend on the PHP runtime, so it's quite not easy to do the Dynamic Snippet Form perfectly.

**Comment by diaafares** [comment]
@billyct yes it worked, but it will be great if Tinkerun returned the error when you hit the Snippet form button, currently if any error happened inside the closure, Tinkerun freezes and you have to Force quit it without knowing whats the problem is.

Try this inside Tinkerun to see whats happen:
```
$users = function() {
    return User::selectRaw('non_exist_column as label, id as value')->limit(3)->get()->toArray();
};

$field_user = [
    'value' => '',
    'label' => 'User',
    'description' => 'Select a user.',
    'type' => 'select',
    'options' => $users, 
];
```

Notice that I added `non_exist_column` in the query, Tinkerun will freeze without display the error to the user, but if you ran the same closure inside terminal tinker you will get the following error message:
```
Illuminate\Database\QueryException with message 'SQLSTATE[42S22]: Column not found: 1054 Unknown column 'non_exist_column' in 'field list'
```

**Comment by billyct** [comment]
@diaafares yes， I think so. I will do it when I am free

**Comment by diaafares** [comment]
@billyct great, can't wait ☺️ thank you so much

**Comment by diaafares** [comment]
Hey @billyct 
Since I upgrade to PHP8 I can't use Dynamic snippet form with something like `User::selectRaw('first_name as label, id as value')->limit(3)->get()->toArray();`

I get the same error as here and when I click `OK`, tinkerun stuck and I have to force quit it:
![image](https://user-images.githubusercontent.com/1811607/134483651-e6bef0a7-1b3b-4431-ae14-05cdac8fea60.png)

I will be grateful if you look at the issue and update Tinkerun to not stuck and show the real error message intead

**Comment by billyct** [comment]
@diaafares 

I have added the error message now, no stuck 😄

<img src="https://user-images.githubusercontent.com/1612364/134673669-3312cee2-40a7-46b6-b45c-0f9b135d68d4.png" width="400"/>

**Comment by diaafares** [comment]
@billyct Thank you so much!
Now Tinkerun show me this error:
```
Error invoking remote method 'parsePhpForm': Error: json_encode((function() {
... return Illuminate\Support\Facades\DB::table('admins')->selectRaw('em ail as label, id as value')->limit(3)->get()->toArray();
... })())
=> "[{"label":"santa.larson@example.com","value":46},{"label":"mschulist@example.org","value":47},{"label":"terry.bud@example.org","value":48}]"
>>>
```

I tried a different table and a different project and I have the same error always, I don't know why 😞

**Comment by billyct** [comment]
@diaafares I think it may be because Laravel/Tinker has no color by default.

https://github.com/laravel/tinker/issues/127

checkout the new version, and let me know if you still have problems.

**Comment by diaafares** [comment]
@billyct 
it is working now, you are amazing
but I have another problem now, after I tested the dynamic snippet form feature, when I hit `command` + `q` to quit Tinkerun, Tinkerun is now freezing and I have to force quit it!
